REGRESSION (iOS 16 Beta 3) Video and audio are broken for WebRTC call after intervention if User enables Siri, receives PSTN call or navigates to Youtube

youennf opened the pull request:
#### 78614b1c3235c130d92d8df8379a9862aa098a64
<pre>
REGRESSION (iOS 16 Beta 3) Video and audio are broken for WebRTC call after intervention if User enables Siri, receives PSTN call or navigates to Youtube
<a href="https://bugs.webkit.org/show_bug.cgi?id=242828">https://bugs.webkit.org/show_bug.cgi?id=242828</a>
rdar://problem/97103335

Reviewed by Eric Carlson.

When there is an audio interruption (Siri or phone call), we sometimes try to uninterrupt by either activating the audio session or starting remote IOs.
This messes up the audio interruption states and we end up not receiving end of interruptions.
We fix the cases where we wrongly try to unterrupt too early.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::tryToSetActive):
tryToSetActive is sometimes called with active=false. In that case, we should just not uninterrupt.
Also tryToSetActive might be called while we are looping over clients to being interruption.
To prevent uninterrupting wrongly, we uninterrupt asynchronously if the AudioSession is active but still interrupted.
(WebCore::AudioSession::beginInterruption):
We sometimes call beginInterruption even if already interrupted.
Bail out early in that case, like we do for endInterruption.
(WebCore::AudioSession::endInterruption):
Add logging to cover the case of endInterruption being called while not interrupted.
Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.mm:
When being interrupted, AVSBDL sometimes throw an exception. This could trigger GPU process crash.
In that case, we catch the exceptions and mark the display layer as failing.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::layersAreInitialized):
Retry initializing layers if it failed.

* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.h:
Export unit() getter.
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::captureUnitHasStopped):
When CoreAudio shared unit is suspended, captureUnitHasStopped is called.
In that case, we should not start the local audio render unit since we are suspending.
Starting to play the local audio render can break Siri or will end up failing in phone call cases.
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.cpp:
Bail out early if we failed creating a LocalSampleBufferDisplayLayer.
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
Make sure to activate the session, if we are about to start capture.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::gpuProcessExited):
Drive-by fix: in case of GPUProcess crash, we need to send the device orientation for the camera video frame orientation to be correctly computed.

Canonical link: <a href="https://commits.webkit.org/252807@main">https://commits.webkit.org/252807@main</a>
</pre>

eric-carlson commented on Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.cpp:
```diff
@@ -77,6 +77,10 @@ void RemoteSampleBufferDisplayLayerManager::createLayer(SampleBufferDisplayLayer
 {
     callOnMainRunLoop([this, protectedThis = Ref { *this }, identifier, hideRootLayer, size, callback = WTFMove(callback)]() mutable {
         auto layer = RemoteSampleBufferDisplayLayer::create(m_connectionToWebProcess, identifier, m_connection.copyRef());
+        if (!layer) {
+            callback({ });
+            return;
+        }
```
It is probably worth logging the error here in a followup.

eric-carlson reviewed (approved):
r=me

youennf committed:
REGRESSION (iOS 16 Beta 3) Video and audio are broken for WebRTC call after intervention if User enables Siri, receives PSTN call or navigates to Youtube
https://bugs.webkit.org/show_bug.cgi?id=242828
rdar://problem/97103335

Reviewed by Eric Carlson.

When there is an audio interruption (Siri or phone call), we sometimes try to uninterrupt by either activating the audio session or starting remote IOs.
This messes up the audio interruption states and we end up not receiving end of interruptions.
We fix the cases where we wrongly try to unterrupt too early.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::tryToSetActive):
tryToSetActive is sometimes called with active=false. In that case, we should just not uninterrupt.
Also tryToSetActive might be called while we are looping over clients to being interruption.
To prevent uninterrupting wrongly, we uninterrupt asynchronously if the AudioSession is active but still interrupted.
(WebCore::AudioSession::beginInterruption):
We sometimes call beginInterruption even if already interrupted.
Bail out early in that case, like we do for endInterruption.
(WebCore::AudioSession::endInterruption):
Add logging to cover the case of endInterruption being called while not interrupted.
Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.mm:
When being interrupted, AVSBDL sometimes throw an exception. This could trigger GPU process crash.
In that case, we catch the exceptions and mark the display layer as failing.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::layersAreInitialized):
Retry initializing layers if it failed.

* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.h:
Export unit() getter.
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::captureUnitHasStopped):
When CoreAudio shared unit is suspended, captureUnitHasStopped is called.
In that case, we should not start the local audio render unit since we are suspending.
Starting to play the local audio render can break Siri or will end up failing in phone call cases.
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.cpp:
Bail out early if we failed creating a LocalSampleBufferDisplayLayer.
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
Make sure to activate the session, if we are about to start capture.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::gpuProcessExited):
Drive-by fix: in case of GPUProcess crash, we need to send the device orientation for the camera video frame orientation to be correctly computed.

Canonical link: https://commits.webkit.org/252807@main
